File: rust/src/bfc.rs
use crate::lexer::{Lexer, Token};

#[derive(PartialEq, Clone, Copy, Debug)]
pub enum OpCode {
    AddB(u8),
    SubB(u8),
    AddP(usize),
    SubP(usize),
    Write,
    Read,
    Jmp(usize),
    Jz(usize),
    End
}

pub struct Compiler {
    token: Token,
    lexer: Lexer,
    currln: usize
}

impl Compiler {
    pub fn from_source(filepath: &str) -> Result<Self, std::io::Error> {
        Ok(Self {
            token: Token::None,
            lexer: Lexer::from_source(filepath)?,
            currln: 0usize
        })
    }

    pub fn compile(&mut self) -> Result<Vec<OpCode>, String> {
        let mut code: Vec<OpCode> = Vec::<OpCode>::new();

        self.token = self.lexer.next_token();
        self.parse_program(&mut code)?;

        Ok(code)
    }

    fn parse_program(&mut self, code: &mut Vec<OpCode>) -> Result<(), String> {
        while self.token != Token::EndOfFile {
            match self.token {
                Token::Plus => self.parse_add_byte(code),
                Token::Minus => self.parse_sub_byte(code),
                Token::ArrowLeft => self.parse_sub_ptr(code),
                Token::ArrowRight => self.parse_add_ptr(code),
                Token::Dot => self.parse_write(code),
                Token::Comma => self.parse_read(code),
                Token::BracketLeft => self.parse_conditional(code)?,
                _ => return Err(format!("unexpected token: {:?}", self.token))
            }
        }

        code.push(OpCode::End);
        Ok(())
    }

    fn parse_add_byte(&mut self, code: &mut Vec<OpCode>) {
        let mut operand: u8 = 0u8;

        while self.token == Token::Plus {
            operand += 1;
            self.token = self.lexer.next_token();
        }

        code.push(OpCode::AddB(operand));
        self.currln += 1;
    }

    fn parse_sub_byte(&mut self, code: &mut Vec<OpCode>) {
        let mut operand: u8 = 0u8;

        while self.token == Token::Minus {
            operand += 1;
            self.token = self.lexer.next_token();
        }

        code.push(OpCode::SubB(operand));
        self.currln += 1;
    }

    fn parse_add_ptr(&mut self, code: &mut Vec<OpCode>) {
        let mut offset: usize = 0usize;

        while self.token == Token::ArrowRight {
            offset += 1;
            self.token = self.lexer.next_token();
        }

        code.push(OpCode::AddP(offset));
        self.currln += 1;
    }

    fn parse_sub_ptr(&mut self, code: &mut Vec<OpCode>) {
        let mut offset: usize = 0usize;

        while self.token == Token::ArrowLeft {
            offset += 1;
            self.token = self.lexer.next_token();
        }

        code.push(OpCode::SubP(offset));
        self.currln += 1;
    }

    fn parse_write(&mut self, code: &mut Vec<OpCode>) {
        code.push(OpCode::Write);
        
        self.token = self.lexer.next_token();
        self.currln += 1;
    }

    fn parse_read(&mut self, code: &mut Vec<OpCode>) {
        code.push(OpCode::Read);

        self.token = self.lexer.next_token();
        self.currln += 1;
    }

    fn parse_conditional(&mut self, code: &mut Vec<OpCode>) -> Result<(), String> {
        let mut braces: Vec<usize> = Vec::<usize>::new();

        braces.push(self.currln);
        self.currln += 1;

        code.push(OpCode::Jz(0usize));

        self.token = self.lexer.next_token();
        while !braces.is_empty() {
            if self.token == Token::EndOfFile {
                return Err(String::from("no matching ']'"));
            }

            match self.token {
                Token::Plus => self.parse_add_byte(code),
                Token::Minus => self.parse_sub_byte(code),
                Token::ArrowLeft => self.parse_sub_ptr(code),
                Token::ArrowRight => self.parse_add_ptr(code),
                Token::Dot => self.parse_write(code),
                Token::Comma => self.parse_read(code),
                Token::BracketLeft => {
                    braces.push(self.currln);
                    self.currln += 1;

                    code.push(OpCode::Jz(0usize));
                    self.token = self.lexer.next_token();
                },
                Token::BracketRight => {
                    if let Some(open) = braces.pop() {
                        self.currln += 1;
                        code[open] = OpCode::Jz(self.currln);
                        code.push(OpCode::Jmp(open));

                        self.token = self.lexer.next_token();
                    } else {
                        return Err(String::from("stack empty"));
                    }
                },
                _ => return Err(format!("unexpected token: {:?}", self.token))
            }
        }

        Ok(())
    }
}

File: python/src/bfc.py

from dataclasses import dataclass
from enum import Enum

from lex import BFLexer
from lex import Token

class OpCode(Enum):
    NONE = 0
    ADDB = 1
    SUBB = 2
    ADDP = 3
    SUBP = 4
    WRITE = 5
    READ = 6
    JZ = 7
    JMP = 8
    END = 9


@dataclass
class ByteCode:
    op: OpCode
    val: int


class BFCompiler:
    __lexer: BFLexer
    __token: Token
    __currln: int

    def __init__(self, filepath: str):
        self.__lexer = BFLexer(filepath)
        self.__token = Token.NONE
        self.__currln = 0

    def compile(self) -> list[ByteCode]:
        code: list[ByteCode] = []
        
        self.__token = self.__lexer.next_token()
        while self.__token != Token.END_OF_FILE:
            match self.__token:
                case Token.PLUS: self.__parse_add_byte(code)
                case Token.MINUS: self.__parse_sub_byte(code)
                case Token.ARROW_LEFT: self.__parse_sub_ptr(code)
                case Token.ARROW_RIGHT: self.__parse_add_ptr(code)
                case Token.DOT: self.__parse_write(code)
                case Token.COMMA: self.__parse_read(code)
                case Token.BRACKET_LEFT: self.__parse_conditional(code)
                case _:
                    print("bfvm: error: unexpected token: \']\'")
                    exit(1)

        code.append(ByteCode(OpCode.END, 0))
        return code

    def __parse_add_byte(self, code: list[ByteCode]):
        val: int = 0

        while self.__token == Token.PLUS:
            val += 1
            self.__token = self.__lexer.next_token()
        
        code.append(ByteCode(OpCode.ADDB, val))
        self.__currln += 1

    def __parse_sub_byte(self, code: list[ByteCode]):
        val: int = 0

        while self.__token == Token.MINUS:
            val += 1
            self.__token = self.__lexer.next_token()
        
        code.append(ByteCode(OpCode.SUBB, val))
        self.__currln += 1

    def __parse_add_ptr(self, code: list[ByteCode]):
        offset: int = 0

        while self.__token == Token.ARROW_RIGHT:
            offset += 1
            self.__token = self.__lexer.next_token()
        
        code.append(ByteCode(OpCode.ADDP, offset))
        self.__currln += 1

    def __parse_sub_ptr(self, code: list[ByteCode]):
        offset: int = 0

        while self.__token == Token.ARROW_LEFT:
            offset += 1
            self.__token = self.__lexer.next_token()
        
        code.append(ByteCode(OpCode.SUBP, offset))
        self.__currln += 1

    def __parse_write(self, code: list[ByteCode]):
        code.append(ByteCode(OpCode.WRITE, 0))
        self.__token = self.__lexer.next_token()
        self.__currln += 1

    def __parse_read(self, code: list[ByteCode]):
        code.append(ByteCode(OpCode.READ, 0))
        self.__token = self.__lexer.next_token()
        self.__currln += 1

    def __parse_conditional(self, code: list[ByteCode]):
        braces: list[int] = []

        braces.append(self.__currln)
        self.__currln += 1

        code.append(ByteCode(OpCode.JZ, 0))
        self.__token = self.__lexer.next_token()
        while braces:
            if self.__token == Token.END_OF_FILE:
                print("bfvm: error: unbalanced braces")
                exit(1)

            match self.__token:
                case Token.PLUS: self.__parse_add_byte(code)
                case Token.MINUS: self.__parse_sub_byte(code)
                case Token.ARROW_LEFT: self.__parse_sub_ptr(code)
                case Token.ARROW_RIGHT: self.__parse_add_ptr(code)
                case Token.DOT: self.__parse_write(code)
                case Token.COMMA: self.__parse_read(code)
                case Token.BRACKET_LEFT:
                    braces.append(self.__currln)
                    self.__currln += 1
                    
                    code.append(ByteCode(OpCode.JZ, 0))
                    self.__token = self.__lexer.next_token()
                case Token.BRACKET_RIGHT:
                    open: int = braces.pop()

                    self.__currln += 1
                    code[open].val = self.__currln

                    code.append(ByteCode(OpCode.JMP, open))
                    self.__token = self.__lexer.next_token()
                case _:
                    print("bfvm: error: unreachable code")
                    exit(1)